pkg/fqdn/re/re.go

```diff
@@ -0,0 +1,92 @@
+// the FQDN subsystem.
+package re
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+
+	lru "github.com/golang/groupcache/lru"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+// CompileRegex provides a convenience wrapper around the LRU. It will either
+// return the cached regex object or compile the pattern into a regex object if
+// the LRU hasn't been initialized (and initializes it) or it doesn't exist in
+// the cache. If it doesn't exist in the cache, it will be added.
+//
+// This function assumes that it's invoked after the LRU has been initialized
+// via InitRegexCompileLRU(), which happens during the DNS proxy
+// initialization, which assumes that the K8s Watcher is not yet synced and no
+// policies are being processed by the system (meaning no policies to parse,
+// i.e. compiling the FQDN rules or handling FQDN selectors). If that doesn't
+// hold true, then we attempt a best-effort to initialize the LRU.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	if RegexCompileLRU == nil {
+		// Try to initialize the LRU here in case the caller never called
+		// InitRegexCompileLRU() explicitly.
+		initOnce.Do(func() {
+			if err := InitRegexCompileLRU(option.Config.FQDNRegexCompileLRUSize); err != nil {
+				log.WithError(err).Warn(
+					"Failed to initialize FQDN regex LRU. Cilium will not cache FQDN regexes. Please report this if you see this message.",
+				)
+			}
+		})
+		// Since this is the case where the LRU is not initialized, we accept
+		// that we won't cache the first result here.
+		return regexp.Compile(p)
+	}
+	RegexCompileLRU.Lock()
+	r, ok := RegexCompileLRU.Get(p)
+	RegexCompileLRU.Unlock()
+	if ok {
+		return r.(*regexp.Regexp), nil
+	}
+	n, err := regexp.Compile(p)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile regex: %w", err)
+	}
+	RegexCompileLRU.Lock()
+	RegexCompileLRU.Add(p, n)
+	RegexCompileLRU.Unlock()
+	return n, nil
+}
+
+// InitRegexCompileLRU initializes the singleton RegexCompileLRU.
+func InitRegexCompileLRU(size int) error {
+	if size <= 0 {
+		return fmt.Errorf("failed to initialize FQDN regex LRU due to invalid size %d", size)
+	}
+	RegexCompileLRU = &regexCompileLRU{
+		Mutex: &lock.Mutex{},
+		Cache: lru.New(size),
+	}
+	return nil
+}
+
+var (
+	// RegexCompileLRU is a singleton instance of the LRU.
+	RegexCompileLRU *regexCompileLRU
+	initOnce        sync.Once
+
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "fqdn/re")
+)
+
+// regexCompileLRU is a shared LRU cache for storing compiled regex objects of
+// FQDN names or patterns, used in CiliumNetworkPolicy or
+// ClusterwideCiliumNetworkPolicy. This is meant to be a singleton that's
+// shared throughout Cilium.
+type regexCompileLRU struct {
+	// The lru package doesn't provide any concurrency guarantees so we must
+	// provide our own locking.
+	*lock.Mutex
```

Comment:
Relatedly, should this use a read/write mutex so that happy path cache hits only use read lock and don't block other readers, while cache miss acquires write lock?

```suggestion
	*lock.RWMutex
```

Reply:
Same answer as here: https://github.com/cilium/cilium/pull/19632#discussion_r862384915